### DISCUSSION TITLE
Multiple Android build-tools folders exist

### DISCUSSION THREAD

**Discussion by marchbold** [airsdk/Adobe-Runtime-Support#997] [Q&A] [answered]
When specifying an Android SDK via the `-platformsdk` option we are seeing the message

```
NOTE: Multiple Android build-tools folders exist, choosing the latest version (29.0.0 rc4) for packaging
```

Just wondering what process you are using for deciding the latest version, as it seems to change for different builds on my machine and is not the latest version I have installed?

eg I have all these installed in `$ANDROID_HOME/build-tools`, as you can see `29.0.0` is not the latest.

```
22.0.0
24.0.0
24.0.3
25.0.0
25.0.2
25.0.3
26.0.0
26.0.1
26.0.2
27.0.0
27.0.2
27.0.3
28.0.0
28.0.0-rc2
28.0.1
28.0.3
29.0.0
29.0.2
29.0.3
31.0.0
```

**Comment by ajwfrost** [comment, marked as answer]
Hi

It basically looks in every (non-hidden) folder under the `build-tools` folder for a file called `source.properties`, and in that file looks for the first line that starts `Pkg.Revision`.
Then there is some (suspicious?) logic to pull out a version code: basically it just removes any non-digit and then parses the outcome as an integer.

So ... yes that's not very clever. `31.0.0` would become 3100, but `29.0.0 rc4` would become 29004 which is somewhat higher...!

I think our options here would be:
1) correct the code so that it actually parses this as a major/minor/etc build number, so we always pick the latest SDK version
2) allow the developer to configure it themselves

For the short term, if you edit the `source.properties` file in the folder you want to use, and e.g. set it to 
`Pkg.Revision=31.0.0 # 00`
then we'll think it's a much higher version number of 310000 and choose that...

cheers

**Reply by marchbold** [reply]
Interesting, so I have a few 29.x versions installed and the contents of that file are :

"29.0.0/source.properties"
``` 
Pkg.UserSrc=false
Pkg.Revision=29.0.0
#Pkg.Revision=29.0.0 rc4
```


"29.0.2/source.properties"
``` 
Pkg.UserSrc=false
Pkg.Revision=29.0.2
#Pkg.Revision=29.0.0 rc4
```

Would your algorithm correctly detect the comment `#` at the beginning? I probably had the release candidate installed at some point, but has since upgraded to the release versions. 

Also I have 31.0.0 installed but it's not using that? Is it too high? 

"31.0.0/source.properties"
```
Pkg.UserSrc=false
Pkg.Revision=31.0.0
#Pkg.Revision=31.0.0 rc5
```

**Reply by marchbold** [reply]
Oh sorry., I just saw your comment on the version becoming 3100, yeah that's definitely not clever...  ;)

You really need to implement a semver comparison there, if you want to do it correctly.

**Reply by ajwfrost** [reply]
Yes it's not the best bit of code I've seen! I think it will ignore the commented-out line purely because it stops looking after it finds the uncommented one...

Won't take a second to make this a bit more robust...

**Reply by marchbold** [reply]
Yeah okay, I was just confused as to where the `29.0.0 rc4` reference in the AIR output is coming from?

**Reply by ajwfrost** [reply]
Ah! just looking again at the code logic, it will only stop looking through the file if it finds a version that's higher than it's "current highest". So if it had looked in 31.0.0 and found that version = 3100; then it looks in your 29.0.0 folder and finds first `Pkg.Revision=29.0.0`, it will work this out as 2900 and ignore it, and then it will actually match with `#Pkg.Revision=29.0.0 rc4` and work that out as 29004, hence choosing this as the version.

Really daft! That version string is only used for the stdout output, so it doesn't matter that it has " rc4" on the end, it will be looking in that 29.0.0 folder. Or maybe 29.0.2. Difficult to tell ...!

**Reply by marchbold** [reply]
Right, okay, will create an issue from this. Doesn't seem to affect anything right now, but thought it was odd.

**Reply by marchbold** [reply]
Transferred to an issue: https://github.com/airsdk/Adobe-Runtime-Support/issues/999
